File: 131.partitionPalindrome.py
class Solution {
    def partition(self, s):
        '''
        :type s: str
        :rtype List[List[str]]
        全局变量或者方法就用self. 如果不是就不加self.
        那么就只是在局部方法作用域内
        '''
        self.res= []
        self.dfs(s, [])
        return self.res

    def dfs(self, s, temp):
        # Return Condition
        if not s:
            self.res.append(temp[:])
            return
        # Backtracking
        for i in xrange(1, len(s)+1):
            if self.isPali(s[:i]):
                temp.append(s[:i])
                self.dfs(s[i:], temp)
                temp.pop()

    def isPali(self, s):
        return s == s[::-1]
}

File: 131.partitionPalindrome.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> out;
        dfs(s, 0, res, out);
        return res;
    }

    void dfs(string s, int start, vector<vector<string>>& res, vector<string>& out) {
        // Return Condition
        if (start == s.size()) {
            res.push_back(out);
            return;
        }
        // Backtracking: 其实跟python代码的思路相同
        for (int i=start; i<s.size(); ++i) {
            if (!isPalindrome(s, start, i)) continue;
            out.push_back(s.substr(start, i-start+1));
            dfs(s, i+1, res, out);
            out.pop_back();
        }
    }

    bool isPalindrome(string s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) return false;
            ++left; --right;
        }
        return true;
    }
}
